codex/blockexchange/swarm.nim

```diff
@@ -0,0 +1,612 @@
+    await self.fillRequests(peer)
+  except CatchableError as err:
+    trace "Error handling block delivery", err = err.msg
+
+proc setupPeer(self: Swarm, peer: PeerId) {.async: (raises: [CancelledError]).} =
+  # If this is an outbound connection, we should already have the peer in our
+  # neighbor set. If it's an inbound connection, it might be the first time we
+  # see this peer.
+  trace "Setting up peer", peer = peer
+
+  self.addNeighbor(peer)
+  # Starts by asking peer for block knowledge.
+  await self.refreshBlockKnowledge(peer)
+
+proc handleBlockKnowledgeRequest(self: Swarm, peer: PeerId) {.async: (raises: []).} =
+  trace "Handling block knowledge request from peer", peer = peer
+  var presenceInfo: seq[BlockPresence]
+  for blockIndex in 0 ..< self.manifest.blocksCount:
+    if self.blocks[blockIndex].completed:
+      presenceInfo.add(
+        BlockPresence(
+          address: init(BlockAddress, self.cid, blockIndex),
+          `type`: BlockPresenceType.Have,
+        )
+      )
+
+  # XXX This will probably be way too expensive to keep sending, even just
+  #   for prototyping
+  if presenceInfo.len > 0:
+    trace "Have block presences to send", count = presenceInfo.len, peer = peer
+  else:
+    trace "No block presences to send, sending empty reply", peer = peer
+    # Sends an easily recognizable bogus block or the reply handler at the
+    # peer will never get triggered.
+    presenceInfo =
+      @[
+        BlockPresence(
+          address: init(BlockAddress, self.cid), `type`: BlockPresenceType.Have
+        )
+      ]
+
+  try:
+    await self.network.request.sendPresence(peer, presenceInfo)
+  except CancelledError:
+    trace "Sending of block presences cancelled", peer = peer
+
+proc handleBlockKnowledgeResponse(
+    self: Swarm, peer: PeerId, presence: seq[BlockPresence]
+) {.async: (raises: []).} =
+  try:
+    if presence.len > 0:
+      trace "Received block knowledge from peer", peer = peer
+      var i = 0
+      for blockPresence in presence:
+        i += 1
+        self.peers[peer].blocks[blockPresence.address.index.int] = true
+
+      trace "Peer has blocks", peer = peer, count = i
+
+      # Learned some potentially new blocks for peer, maybe we can push
+      # some more requests.
+      await self.fillRequests(peer)
+    else:
+      trace "Peer reported no block knowledge", peer = peer
+  except KeyError:
+    error "Cannot update block presence for peer", peer = peer
+    return
+  except CancelledError:
+    trace "Sending of block requests cancelled", peer = peer
+  finally:
+    try:
+      if not self.peers[peer].pendingBlockKnowledgeRequest:
+        trace "Illegal state for pending block knowledge request (already cleared)",
+          peer = peer
+      self.peers[peer].pendingBlockKnowledgeRequest = false
+    except KeyError:
+      trace "Error updating pending block knowledge request state (peer dropped)",
+        peer = peer
+
+proc handleBlockRequest(
```

Comment:
I know this is an experiment, but we should be lazy with handling requests, so I would still put this behind some throttling mechanism, such as the asyncqueue.